syntax = "proto3";

package protos;

service Benchmarks{
 rpc track_objects(Request) returns (DetectionTrackingResponse) {}
 rpc detect_objects(Request) returns (DetectionTrackingResponse) {}
 rpc speech_to_text(Request) returns (PocketSphinxResponse) {}
 rpc align_speech_text(AudioTextRequest) returns (PocketSphinxResponse) {}
 rpc start_memory_tracing(EmptyProto) returns (EmptyProto) {}
 rpc get_cpu_trace(EmptyProto) returns (CPUTrace) {}
 rpc get_memory_usage(EmptyProto) returns (MemoryTrace) {}
 rpc inject_fault(FaultRequest) returns (EmptyProto) {}
 rpc get_fault_injection_status(EmptyProto) returns (FaultInjectionStatus) {}
}

message Request{
 string image = 1;
 int64 frame_id = 2;
 int64 request_time_ms = 3;
}

message AudioTextRequest{
 string audio = 1;
 string text_input = 2;
 int64 frame_id = 3;
 int64 request_time_ms = 4;
}

message FaultRequest{
 string fault_command = 1;
 string fault_config = 2;
 int32 timeout = 3;
}

message DetectionTrackingResponse{
 int64 frame_id = 1;
 int64 request_time_ms = 2;
 int64 request_received_time_ms = 3;
 int64 response_time_ms = 4;
 int64 response_received_time_ms = 5;
 repeated DetectedTrackedObject detected_objects = 6;
}

message PocketSphinxResponse{
 int64 frame_id = 1;
 int64 request_time_ms = 2;
 int64 request_received_time_ms = 3;
 int64 response_time_ms = 4;
 int64 response_received_time_ms = 5;
 string conversion_result = 6;
}

message DetectedTrackedObject {
  int32 track_id = 1;
  string clazz = 2;
  int32 x_min = 3;
  int32 x_max = 4;
  int32 y_min = 5;
  int32 y_max = 6;
}

message CPUTrace {
  float cpu_load = 1;
}

message MemoryTrace {
  float current_memory_mb = 1;
  float peak_memory_mb = 2;
}

message FaultInjectionStatus {
  bool is_finished = 1;
}

message EmptyProto {}
